packages/builders/src/messages/MessageReference.ts

```diff
@@ -0,0 +1,98 @@
+import type { JSONEncodable } from '@discordjs/util';
+import type { APIMessageReference, MessageReferenceType, Snowflake } from 'discord-api-types/v10';
+import { validate } from '../util/validation.js';
+import { messageReferencePredicate } from './Assertions.js';
+
+/**
+ * A builder that creates API-compatible JSON data for message references.
+ */
+export class MessageReferenceBuilder implements JSONEncodable<APIMessageReference> {
+	private readonly data: Partial<APIMessageReference>;
+
+	/**
+	 * Creates new allowed mention builder from API data.
+	 *
+	 * @param data - The API data to create this attachment with
+	 */
+	public constructor(data: Partial<APIMessageReference> = {}) {
+		this.data = structuredClone(data);
+	}
+
+	/**
+	 * Sets the types of message reference this represents
+	 *
+	 * @param type - The type of message reference
+	 */
+	public setType(type: MessageReferenceType): this {
+		this.data.type = type;
+		return this;
+	}
+
+	/**
+	 * Clear the type of message reference this represents
+	 */
+	public clearType(): this {
+		this.data.type = undefined;
+		return this;
+	}
+
+	/**
+	 * Sets the id of the message being referenced
+	 *
+	 * @param messageId - The id of the message being referenced
+	 */
+	public setMessageId(messageId: Snowflake): this {
+		this.data.message_id = messageId;
+		return this;
+	}
+
+	/**
+	 * Clear the id of the message being referenced
+	 */
+	public clearMessageId(): this {
```

Comment:
Weird to have a clear message id when it is always needed iirc

Reply:
According to your types, it's optional, hence why I added it.

Reply:
Its optional on receive, not send